File: examples/chicken/kelvin-helmholtz/kh.py
# kh.py
# A simple 2D Kelvin-Helmholtz instability.
# PJ 2022-10-09 Adapted from Lachlan Whyborn's Eilmer4 example.
#
import math
config.title = "Kelvin-Helmholtz instability."
print(config.title)
#
H = 0.010 # z-direction layer thickness in metres
L = 1.000 # x,y-direction domain size in metres
#
vol0 = TFIVolume(p000=Vector3(0,0,0), p100=Vector3(L,0,0),
                 p110=Vector3(L,L,0), p010=Vector3(0,L,0),
                 p001=Vector3(0,0,H), p101=Vector3(L,0,H),
                 p111=Vector3(L,L,H), p011=Vector3(0,L,H))
N = 256
grd0 = StructuredGrid(pvolume=vol0, niv=N+1, njv=N+1, nkv=3)
#
# Flow conditions
U1 = 50.0; U2 = -50.0; U_m = (U1-U2)/2
rho1 = 1.0; rho2 = 2.0; rho_m = (rho1-rho2)/2
p_inf = 1.0e5
Rgas = 287.1; g = 1.4
Ls = 0.025

def initial_flow(x, y, z):
    """
    User-defined function for the initial flow state works in physical space.
    """
    if y < L/4:
        rho = rho1 - rho_m * math.exp((y-L/4)/Ls)
        U = U1 - U_m * math.exp((y-L/4)/Ls)
    elif y >= L/4 and y < L/2:
        rho = rho2 + rho_m * math.exp((-y+L/4)/Ls)
        U = U2 + U_m * math.exp((-y+L/4)/Ls)
    elif y >= L/2 and y < 3*L/4:
        rho = rho2 + rho_m * math.exp(-(3*L/4-y)/Ls)
        U = U2 + U_m * math.exp(-(3*L/4-y)/Ls)
    else:
        rho = rho1 - rho_m * math.exp(-(y-3*L/4)/Ls)
        U = U1 - U_m * math.exp(-(y-3*L/4)/Ls)
    #
    T = p_inf/(Rgas*rho)
    vy_perturb = 1.0 * math.sin(4*x*math.pi)
    return FlowState(p=p_inf, velx=U, vely=vy_perturb, T=T)
#
blk0 = FluidBlock(grid=grd0, initialState=initial_flow,
                  bcs={'iminus':ExchangeBC(),'iplus':ExchangeBC(),
                       'jminus':ExchangeBC(),'jplus':ExchangeBC()})
#
ref_time = L/abs(U1-U2)
config.max_time = 2.5 * ref_time
config.max_step = 1000000
add_dt_plot(0.0, config.max_time/25)

File: examples/eilmer/2D/kelvin-helmholtz/kh.lua
-- A simple 2D Kelvin-Helmholtz instability

config.title = "KH"
config.dimensions = 2

setGasModel('ideal-air.lua')

-- Configure the domain
L = 1
a = Vector3:new{x = 0, y = 0}
b = Vector3:new{x = L, y = 0}
c = Vector3:new{x = 0, y = L}
d = Vector3:new{x = L, y = L}

south = Line:new{p0 = a, p1 = b}
west = Line:new{p0 = a, p1 = c}
north = Line:new{p0 = c, p1 = d}
east = Line:new{p0 = b, p1 = d}

psurf = makePatch{south = south, west = west, north = north, east = east}

N = 256
sgrid = StructuredGrid:new{psurface = psurf, niv = N + 1, njv = N + 1}

-- The flow conditions
U1 = 50; U2 = -50; U_m = (U1 - U2) / 2
rho1 = 1; rho2 = 2; rho_m = (rho1 - rho2) / 2
p_inf = 1e5
Rgas = 287.15; gamma = 1.4
Ls = 0.025

function KH_flow(x, y)
    if y < (L / 4) then
        rho = rho1 - rho_m * math.exp((y - L / 4) / Ls)
        U = U1 - U_m * math.exp((y - L / 4) / Ls)
    elseif (y >= (L / 4) and y < (L / 2)) then
        rho = rho2 + rho_m * math.exp((-y + L / 4) / Ls)
        U = U2 + U_m * math.exp((-y + L / 4) / Ls)
    elseif (y >= (L / 2) and y < (3 * L / 4)) then
        rho = rho2 + rho_m * math.exp(-(3 * L / 4 - y) / Ls)
        U = U2 + U_m * math.exp(-(3 * L / 4 - y) / Ls)
    else
        rho = rho1 - rho_m * math.exp(-(y - 3 * L / 4) / Ls)
        U = U1 - U_m * math.exp(-(y - 3 * L / 4) / Ls)
    end

    T = p_inf / (Rgas * rho)
    vy_perturb = 1 * math.sin(4 * math.pi * x)
    return FlowState:new{T = T, p = p_inf, velx = U, vely = vy_perturb}
end

nib = 4; njb = 4
FBA = FBArray:new{grid = sgrid, fillCondition = KH_flow, bcList = {south = OutFlowBC_SimpleExtrapolate:new{}, north = OutFlowBC_SimpleExtrapolate:new{}}, nib = nib, njb = njb}

for j = 1, njb do
    connectBlocks(FBA.blockArray[1][j], "west", FBA.blockArray[nib][j], "east")
end
for i = 1, nib do
    connectBlocks(FBA.blockArray[i][1], "south", FBA.blockArray[i][njb], "north")
end

identifyBlockConnections()
mpiDistributeBlocks{ntasks = nib * njb, dist = 'load-balance'}

ref_time = L / math.abs(U1 - U2)
config.cfl_value = 0.5
config.viscous = false
config.gasdynamic_update_scheme = 'classic-rk3'
-- config.flux_calculator = 'ldfss0'
config.flux_calculator = 'ausmdv'
-- config.apply_limiter = false
config.extrema_clipping = false
config.max_time = 2.5 * ref_time
config.dt_plot = config.max_time / 25
config.max_step = 1000000
config.new_flow_format = true
config.flow_format = "eilmer4text"

